File: src/2025-02-27-deepclaude37-rust/javascript/exercises/practice/forth/forth.js
//
// This is only a SKELETON file for the 'Forth' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

export class Forth {
  constructor() {
    this._stack = [];
    this.customWords = new Map();
  }

  evaluate(input) {
    const tokens = this.tokenize(input);
    this.processTokens(tokens);
  }

  tokenize(input) {
    return input.split(/\s+/g)
      .filter(t => t !== '')
      .map(t => {
        if (/^-?\d+$/.test(t)) {
          return t;
        } else {
          return t.toUpperCase();
        }
      });
  }

  processTokens(tokens) {
    let i = 0;
    while (i < tokens.length) {
      const token = tokens[i];
      if (token === ':') {
        i = this.parseDefinition(tokens, i);
      } else {
        this.handleToken(token);
        i++;
      }
    }
  }

  parseDefinition(tokens, i) {
    if (i + 1 >= tokens.length) {
      throw new Error('Invalid definition');
    }
    const name = tokens[i + 1];
    if (/^-?\d+$/.test(name)) {
      throw new Error('Cannot redefine numbers');
    }
    let j = i + 2;
    const definition = [];
    while (j < tokens.length && tokens[j] !== ';') {
      definition.push(tokens[j]);
      j++;
    }
    if (j >= tokens.length || tokens[j] !== ';') {
      throw new Error('Unclosed definition');
    }
    this.customWords.set(name.toUpperCase(), definition);
    return j + 1;
  }

  handleToken(token) {
    if (/^-?\d+$/.test(token)) {
      this._stack.push(parseInt(token, 10));
    } else {
      if (this.customWords.has(token)) {
        const definition = this.customWords.get(token);
        this.processTokens(definition);
      } else {
        this.executeBuiltIn(token);
      }
    }
  }

  executeBuiltIn(token) {
    switch (token) {
      case '+':
        this.ensureStackSize(2);
        const a1 = this._stack.pop();
        const b1 = this._stack.pop();
        this._stack.push(b1 + a1);
        break;
      case '-':
        this.ensureStackSize(2);
        const a2 = this._stack.pop();
        const b2 = this._stack.pop();
        this._stack.push(b2 - a2);
        break;
      case '*':
        this.ensureStackSize(2);
        const a3 = this._stack.pop();
        const b3 = this._stack.pop();
        this._stack.push(b3 * a3);
        break;
      case '/':
        this.ensureStackSize(2);
        const a4 = this._stack.pop();
        if (a4 === 0) {
          throw new Error('Division by zero');
        }
        const b4 = this._stack.pop();
        this._stack.push(Math.trunc(b4 / a4));
        break;
      case 'DUP':
        this.ensureStackSize(1);
        const top = this._stack[this._stack.length - 1];
        this._stack.push(top);
        break;
      case 'DROP':
        this.ensureStackSize(1);
        this._stack.pop();
        break;
      case 'SWAP':
        this.ensureStackSize(2);
        const a5 = this._stack.pop();
        const b5 = this._stack.pop();
        this._stack.push(a5);
        this._stack.push(b5);
        break;
      case 'OVER':
        this.ensureStackSize(2);
        const second = this._stack[this._stack.length - 2];
        this._stack.push(second);
        break;
      default:
        throw new Error(`Unknown word: ${token}`);
    }
  }

  ensureStackSize(size) {
    if (this._stack.length < size) {
      throw new Error('Stack underflow');
    }
  }

  get stack() {
    return [...this._stack];
  }
}

File: src/2025-02-27-deepclaude37-rust/python/exercises/practice/forth/forth.py
class StackUnderflowError(Exception):
    """Exception raised when Stack is not full.
       message: explanation of the error.
    """
    def __init__(self, message):
        self.message = message
        super().__init__(message)


def evaluate(input_data):
    stack = []
    custom_words = {}

    def tokenize(line):
        tokens = []
        for token in line.split():
            tokens.append(token.lower() if not is_number(token) else token)
        return tokens

    def is_number(s):
        try:
            int(s)
            return True
        except ValueError:
            return False

    def process_tokens(tokens, stack, custom_words):
        i = 0
        while i < len(tokens):
            token = tokens[i]
            
            # Handle word definition
            if token == ':':
                if i + 1 >= len(tokens):
                    raise ValueError("Invalid word definition")
                word_name = tokens[i + 1]
                
                # Word names cannot be numbers
                if is_number(word_name):
                    raise ValueError("illegal operation")
                
                # Collect definition until semicolon
                j = i + 2
                while j < len(tokens) and tokens[j] != ';':
                    j += 1
                
                if j >= len(tokens) or tokens[j] != ';':
                    raise ValueError("Missing semicolon")
                
                # Store definition
                definition = tokens[i+2:j]
                
                # Recursively expand custom words in the definition at definition time
                def expand_definition(tokens, words):
                    expanded = []
                    for token in tokens:
                        if token in words:
                            expanded.extend(expand_definition(words[token], words))
                        else:
                            expanded.append(token)
                    return expanded
                
                expanded_def = expand_definition(definition, custom_words)
                custom_words[word_name] = expanded_def
                i = j + 1
            
            # Process regular token
            else:
                if is_number(token):
                    stack.append(int(token))
                elif token in custom_words:
                    # Instead of recursive processing, insert the definition directly
                    definition = custom_words[token]
                    # Process each token in the definition
                    for def_token in definition:
                        if is_number(def_token):
                            stack.append(int(def_token))
                        else:
                            # Handle built-in operations
                            if def_token == '+':
                                if len(stack) < 2:
                                    raise StackUnderflowError("Insufficient number of items in stack")
                                a, b = stack.pop(), stack.pop()
                                stack.append(b + a)
                            elif def_token == '-':
                                if len(stack) < 2:
                                    raise StackUnderflowError("Insufficient number of items in stack")
                                a, b = stack.pop(), stack.pop()
                                stack.append(b - a)
                            elif def_token == '*':
                                if len(stack) < 2:
                                    raise StackUnderflowError("Insufficient number of items in stack")
                                a, b = stack.pop(), stack.pop()
                                stack.append(b * a)
                            elif def_token == '/':
                                if len(stack) < 2:
                                    raise StackUnderflowError("Insufficient number of items in stack")
                                a, b = stack.pop(), stack.pop()
                                if a == 0:
                                    raise ZeroDivisionError("divide by zero")
                                stack.append(b // a)
                            elif def_token == 'dup':
                                if len(stack) < 1:
                                    raise StackUnderflowError("Insufficient number of items in stack")
                                stack.append(stack[-1])
                            elif def_token == 'drop':
                                if len(stack) < 1:
                                    raise StackUnderflowError("Insufficient number of items in stack")
                                stack.pop()
                            elif def_token == 'swap':
                                if len(stack) < 2:
                                    raise StackUnderflowError("Insufficient number of items in stack")
                                a, b = stack.pop(), stack.pop()
                                stack.append(a)
                                stack.append(b)
                            elif def_token == 'over':
                                if len(stack) < 2:
                                    raise StackUnderflowError("Insufficient number of items in stack")
                                stack.append(stack[-2])
                            else:
                                raise ValueError("undefined operation")
                else:
                    # Process built-in operations
                    if token == '+':
                        if len(stack) < 2:
                            raise StackUnderflowError("Insufficient number of items in stack")
                        a, b = stack.pop(), stack.pop()
                        stack.append(b + a)
                    elif token == '-':
                        if len(stack) < 2:
                            raise StackUnderflowError("Insufficient number of items in stack")
                        a, b = stack.pop(), stack.pop()
                        stack.append(b - a)
                    elif token == '*':
                        if len(stack) < 2:
                            raise StackUnderflowError("Insufficient number of items in stack")
                        a, b = stack.pop(), stack.pop()
                        stack.append(b * a)
                    elif token == '/':
                        if len(stack) < 2:
                            raise StackUnderflowError("Insufficient number of items in stack")
                        a, b = stack.pop(), stack.pop()
                        if a == 0:
                            raise ZeroDivisionError("divide by zero")
                        stack.append(b // a)
                    elif token == 'dup':
                        if len(stack) < 1:
                            raise StackUnderflowError("Insufficient number of items in stack")
                        stack.append(stack[-1])
                    elif token == 'drop':
                        if len(stack) < 1:
                            raise StackUnderflowError("Insufficient number of items in stack")
                        stack.pop()
                    elif token == 'swap':
                        if len(stack) < 2:
                            raise StackUnderflowError("Insufficient number of items in stack")
                        a, b = stack.pop(), stack.pop()
                        stack.append(a)
                        stack.append(b)
                    elif token == 'over':
                        if len(stack) < 2:
                            raise StackUnderflowError("Insufficient number of items in stack")
                        stack.append(stack[-2])
                    else:
                        raise ValueError("undefined operation")
                i += 1

    # Process each line in the input
    for line in input_data:
        if line.strip():  # Skip empty lines
            tokens = tokenize(line)
            process_tokens(tokens, stack, custom_words)

    return stack
